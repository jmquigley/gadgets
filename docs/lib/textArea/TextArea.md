<a name="module_TextArea"></a>

## TextArea
A multiline text editing component.  It is a a contenteditable div.  As text
is added to the component the onUpdate callback is executed to pass the
contents that have chnaged.

## Screen:
<img src="https://github.com/jmquigley/gadgets/blob/master/images/textarea.png" width="60%" />

## Examples:

```javascript
import {TextAra} from "gadgets";

<TextArea
    maxRows={10}
    onUpdate={callback}
    value={initial text}
/>
```

## API
#### Events
- `onUpdate(text: string)` - Invoked when the text within the component is changed.
The current text in the component is passed to the callback.

#### Styles
- `ui-textarea` - Placed on the `<div>` component that wraps the text component.

#### Properties
- `rows=5 {number}` - The number of lines displayed within the component.
- `updateDelay=150 {number}` - The onUpdate function is called when the text is
updated.  This is a debounce delay to rate limit how often this function is
called when input into the control is fast.  The number is milliseconds.
- `value=null {string}` - a string value that will overwrite the contents of
the current component.

