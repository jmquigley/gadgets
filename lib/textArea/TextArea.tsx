/**
 * A multiline text editing component.  It is a a contenteditable div.  As text
 * is added to the component the onUpdate callback is executed to pass the
 * contents that have chnaged.
 *
 * ## Screen:
 * <img src="https://github.com/jmquigley/gadgets/blob/master/images/textarea.png" width="80%" />
 *
 * ## Examples:
 *
 * ```javascript
 * import {TextAra} from "gadgets";
 *
 * <TextArea
 *     maxRows={10}
 *     onUpdate={callback}
 *     value={initial text}
 * />
 * ```
 *
 * ## API
 * #### Events
 * - `onUpdate(text: string)` - Invoked when the text within the component is changed.
 * The current text in the component is passed to the callback.
 *
 * #### Styles
 * - `ui-textarea` - Placed on the `<div>` component that wraps the text component.
 *
 * #### Properties
 * - `rows=5 {number}` - The number of lines displayed within the component.
 * - `updateDelay=150 {number}` - The onUpdate function is called when the text is
 * updated.  This is a debounce delay to rate limit how often this function is
 * called when input into the control is fast.  The number is milliseconds.
 * - `value=null {string}` - a string value that will overwrite the contents of
 * the current component.
 *
 * @module TextArea
 */

// const debug = require("debug")("gadgets.TextArea");

import autobind from "autobind-decorator";
import {debounce} from "lodash";
import * as React from "react";
import styled from "styled-components";
import {unitToNumber} from "util.calc";
import {splitNL} from "util.string";
import {nilEvent} from "util.toolbox";
import {
	BaseComponent,
	BaseProps,
	BaseState,
	disabled,
	fontStyle,
	getDefaultBaseProps,
	getDefaultBaseState,
	invisible,
	Wrapper
} from "../shared";

export interface TextAreaProps extends BaseProps {
	onChange?: (e: React.ChangeEvent<HTMLDivElement>) => void;
	onUpdate?: (text: string) => void;
	rows?: number;
	updateDelay?: number;
	value?: string;
}

export function getDefaultTextAreaProps(): TextAreaProps {
	return {
		...getDefaultBaseProps(),
		obj: "TextArea",
		onChange: nilEvent,
		onUpdate: nilEvent,
		padding: "4px",
		rows: 5,
		updateDelay: 150,
		value: null
	};
}

export interface TextAreaState extends BaseState {
	value: string;
}

export function getDefaultTextAreaState(): TextAreaState {
	return {
		...getDefaultBaseState(),
		value: ""
	};
}

const TextAreaView: any = styled.div`
	-webkit-transition: all ${(props: TextAreaProps) =>
		props.theme.transitionDelay} ease-in-out;
	border: solid 1px ${(props: TextAreaProps) => props.theme.borderColor};
	height: ${(props: TextAreaProps) => props.maxHeight};
	max-height: ${(props: TextAreaProps) => props.maxHeight};
	outline: none;
	overflow: auto;
	padding: ${(props: TextAreaProps) => props.padding || "4px"};
	width: 100%;

	&:focus {
	    border: 1px solid ${(props: TextAreaProps) => props.theme.outlineColor};
		box-shadow: 0 0 5px ${(props: TextAreaProps) => props.theme.outlineColor};
	}

	${(props: TextAreaProps) => disabled(props)}
	${(props: TextAreaProps) => invisible(props)}
	${(props: TextAreaProps) => props.sizing && fontStyle[props.sizing]}
`;

export class TextArea extends BaseComponent<TextAreaProps, TextAreaState> {
	public static readonly defaultProps: TextAreaProps = getDefaultTextAreaProps();
	private _container: HTMLDivElement = null;
	private _debounceUpdate: any = null;

	constructor(props: TextAreaProps) {
		super(props, "ui-textarea", TextArea.defaultProps.style);

		this._debounceUpdate = debounce(
			this.props.onUpdate,
			this.props.updateDelay
		);

		this.state = {
			...getDefaultTextAreaState(),
			value: this.props.value
		};
	}

	@autobind
	private handleChange(e: React.ChangeEvent<HTMLDivElement>) {
		const target: HTMLDivElement = e.target as HTMLDivElement;
		const value: string = target.innerText;

		this.props.onChange(e);
		this._debounceUpdate(value);
	}

	@autobind
	private handleRef(ref: any) {
		this._container = ref;
	}

	public static getDerivedStateFromProps(
		props: TextAreaProps,
		state: TextAreaState
	) {
		if (props.value && props.value !== state.value) {
			const newState: TextAreaState = {
				...state,
				value: props.value
			};

			return super.getDerivedStateFromProps(props, newState, true);
		}

		return null;
	}

	public componentDidMount() {
		this.updateContent(this.state.value);
	}

	public componentDidUpdate(
		_prevProps: TextAreaProps,
		prevState: TextAreaState
	) {
		if (prevState.value !== this.state.value) {
			this.updateContent(this.state.value);
		}
	}

	private updateContent(text: string) {
		const lines: string[] = splitNL(text);
		this._container.innerHTML = lines
			.map((it: string) => {
				return `<div>${it}</div>`;
			})
			.join("");

		this._debounceUpdate(text);
	}

	public render() {
		this.updateClassName();

		const {onUpdate, ...props} = this.props;

		const totalRows = props.rows < 1 ? 1 : props.rows;
		const padding: number = unitToNumber(props.padding);
		const lineHeight: number = unitToNumber(
			BaseComponent.lineHeightPX(props.sizing)
		);
		const maxHeight = lineHeight * totalRows + padding * 2;

		return (
			<Wrapper {...props}>
				<TextAreaView
					{...props}
					className={this.className}
					contentEditable={true}
					maxHeight={`${maxHeight}px`}
					onInput={this.handleChange}
					ref={this.handleRef}
					style={this.state.style}
					suppressContentEditableWarning
				/>
			</Wrapper>
		);
	}
}

export default TextArea;
