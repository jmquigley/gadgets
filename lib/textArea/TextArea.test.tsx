import {mount, shallow} from "enzyme";
import assert from "power-assert";
import * as React from "react";
import TextArea from "./TextArea";

test("Create a simple TextArea widget", () => {
	const ctl = mount(<TextArea rows={10} value='some text' />);

	assert(ctl);
	expect(ctl).toMatchSnapshot();
});
