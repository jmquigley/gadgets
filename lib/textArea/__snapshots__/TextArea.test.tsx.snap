// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a simple TextArea widget 1`] = `
<TextArea
  backgroundColor={null}
  border={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  hidden={false}
  id={null}
  left={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noborder={false}
  noedit={false}
  nohover={false}
  nopropagation={false}
  notheme={false}
  notooltip={false}
  onChange={[Function]}
  onUpdate={[Function]}
  padding="4px"
  ripple={false}
  rows={10}
  selected={false}
  sizing="normal"
  style={Object {}}
  testing={true}
  theme={
    Object {
      "backgroundColor": "#ffffff",
      "borderColor": "#c0c0c0",
      "buttonColor": "#666666",
      "chevronColor": "#cccccc",
      "color": "#000000",
      "fontFamily": "OpenSans, Georgia, serif",
      "headerBackgroundColor": "#004358",
      "headerForegroundColor": "#ffffff",
      "headerHoverColor": "#ffe11a",
      "hoverColor": "#c0c0c0",
      "inputBorderColor": "#c0c0c0",
      "itemAlternatingColor": "#e6e6e6",
      "itemHoverColor": "#bedb39",
      "outlineColor": "#51cbee",
      "searchFocus": "#51cbee",
      "searchMatch": "#fd7400",
      "selectedBackgroundColor": "#1f8a70",
      "selectedForegroundColor": "#ffe11a",
      "titleBarBackgroundColor": "#004358",
      "titleBarForegroundColor": "#ffe11a",
      "tooltipBackgroundColor": "#708090",
      "tooltipForegroundColor": "#ffffff",
      "transitionDelay": "0.33s",
    }
  }
  tooltip={null}
  top={null}
  updateDelay={150}
  value="some text"
  visible={true}
  width={null}
  xcss={null}
/>
`;
